src/Symfony/Bundle/FrameworkBundle/Secrets/SodiumVault.php

```diff
@@ -0,0 +1,164 @@
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class SodiumVault
+{
+    private $encryptionKey;
+    private $decryptionKey;
+    private $secretsDir;
+    private $encryptionKeyName;
+    private $decryptionKeyName;
+
+    public function __construct(string $secretsDir)
+    {
+        if (!\function_exists('sodium_crypto_box_seal')) {
+            throw new \LogicException('The "sodium" PHP extension is required to deal with secrets. Alternatively, try running "composer require paragonie/sodium_compat" if you cannot enable the extension."');
+        }
+
+        if (!is_dir($secretsDir) && !@mkdir($secretsDir, 0777, true) && !is_dir($secretsDir)) {
+            throw new \RuntimeException(sprintf('Unable to create the secrets directory (%s)', $secretsDir));
+        }
+
+        $this->secretsDir = rtrim($secretsDir, '/'.\DIRECTORY_SEPARATOR).\DIRECTORY_SEPARATOR;
+
+        $baseName = basename($this->secretsDir);
+        $this->encryptionKeyName = $baseName.'.sodium.public.key';
+        $this->decryptionKeyName = $baseName.'.sodium.secret.key';
+    }
+
+    public function generateKeys(bool $override = false): bool
+    {
+        if (!$override && (file_exists($this->secretsDir.$this->encryptionKeyName) || file_exists($this->secretsDir.$this->decryptionKeyName))) {
+            return false;
+        }
+
+        $this->decryptionKey = sodium_crypto_box_keypair();
+        $this->encryptionKey = sodium_crypto_box_publickey($this->decryptionKey);
+
+        $this->export($this->encryptionKeyName, $this->encryptionKey);
+        $this->export($this->decryptionKeyName, $this->decryptionKey);
+
+        return true;
+    }
+
+    public function seal(string $name, string $secret): void
+    {
+        $this->loadKeys($name);
+        $this->export($name.'.sodium', sodium_crypto_box_seal($secret, $this->encryptionKey));
+    }
+
+    public function reveal(string $name): ?string
+    {
+        $this->loadKeys($name);
+
+        if (!file_exists($file = $this->secretsDir.$name.'.sodium')) {
+            return null;
+        }
+
+        if (!\is_string($sealedSecret = include $file)) {
+            throw new \LogicException(sprintf('Cannot load secret "%s".', $name));
+        }
+
+        return sodium_crypto_box_seal_open($sealedSecret, $this->decryptionKey);
+    }
+
+    public function list(bool $reveal = false): array
+    {
+        $this->loadKeys();
+        $secrets = [];
+
+        foreach (scandir($this->secretsDir) as $name) {
+            if (0 === substr_compare($name, '.sodium', -\strlen('.sodium'))) {
+                $name = substr($name, 0, -\strlen('.sodium'));
+                $secrets[$name] = $reveal ? $this->reveal($name) : '********';
+            }
+        }
+
+        return $secrets;
+    }
+
+    public function zero(string &...$sensitiveData): void
+    {
+        foreach ($sensitiveData as &$data) {
+            if (\extension_loaded('sodium')) {
```

Comment:
does it worse moving the if outside the loop?

Reply:
`\extension_loaded` is optimized out by opcache